storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -55,6 +113,2292 @@ public class UnifiedLog {
+                        // it may not be true following a message format version bump as the epoch will not be available for
+                        // log entries written in the older format.
+                        Optional<EpochEntry> earliestEpochEntry = leaderEpochCache.earliestEntry();
+                        Optional<Integer> epochOpt = (earliestEpochEntry.isPresent() && earliestEpochEntry.get().startOffset <= logStartOffset)
+                            ? Optional.of(earliestEpochEntry.get().epoch)
+                            : Optional.empty();
+
+                        return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, logStartOffset, epochOpt));
+                    } else if (targetTimestamp == ListOffsetsRequest.EARLIEST_LOCAL_TIMESTAMP) {
+                        long curLocalLogStartOffset = localLogStartOffset();
+
+                        OptionalInt epochForOffset = leaderEpochCache.epochForOffset(curLocalLogStartOffset);
+                        Optional<Integer> epochResult = epochForOffset.isPresent()
+                            ? Optional.of(epochForOffset.getAsInt())
+                            : Optional.empty();
+
+                        return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, curLocalLogStartOffset, epochResult));
+                    } else if (targetTimestamp == ListOffsetsRequest.LATEST_TIMESTAMP) {
+                        return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, logEndOffset(), leaderEpochCache.latestEpoch()));
+                    } else if (targetTimestamp == ListOffsetsRequest.LATEST_TIERED_TIMESTAMP) {
+                        if (remoteLogEnabled()) {
+                            long curHighestRemoteOffset = highestOffsetInRemoteStorage();
+                            OptionalInt epochOpt = leaderEpochCache.epochForOffset(curHighestRemoteOffset);
+                            Optional<Integer> epochResult = epochOpt.isPresent()
+                                ? Optional.of(epochOpt.getAsInt())
+                                : curHighestRemoteOffset == -1
+                                    ? Optional.of(RecordBatch.NO_PARTITION_LEADER_EPOCH)
+                                    : Optional.empty();
+                            return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, curHighestRemoteOffset, epochResult));
+                        } else {
+                            return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, -1L, Optional.of(-1)));
+                        }
+                    } else if (targetTimestamp == ListOffsetsRequest.MAX_TIMESTAMP) {
+                        // Cache to avoid race conditions.
+                        List<LogSegment> segments = List.copyOf(logSegments());
```

Comment:
This is an existing issue. We are iterating `localLog.segments().values()` when it can be updated. So the behavior is undefined. It's safer to create a copy while holding the lock.

Reply:
Right. So we probably also want to do it in `searchOffsetInLocalLog()` too.